### PR TITLE
feat(VILLB-6): create send login otp email controller and test cases.

### PR DESCRIPTION
**Pull request by katafalque**

### CODE REVIEW

File: src/login/login.module.ts [hunk 1]
```diff
@@ -5,10 +5,12 @@ import { Module } from "@nestjs/common";
 import { OtpModule } from "../otp/otp.module";
 import { TokenModule } from "../token/token.module";
 import { UserModule } from "../user/user.module";
+import { SendLoginOtpEmailController } from "./send-otp-email-login.controller";
+import {EventEmitterModule} from '@nestjs/event-emitter';
 
 @Module({
-  imports: [TokenModule, UserModule, OtpModule],
-  controllers: [LoginController, LoginOtpEmailController],
+  imports: [TokenModule, UserModule, OtpModule, EventEmitterModule.forRoot()],
```
Comment (ismetkizgin) [inline src/login/login.module.ts:12]: `EventEmitterModule.forRoot()` is defined in the `app` module. Let's remove this definition.

File: src/login/send-otp-email-login.controller.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    private readonly eventEmitter: EventEmitter2
+  ) {}
+
+  @Post("send-login-otp-email")
+  async sendLoginOtpEmailAsync(@Body() sendLoginOtpEmailRequest: SendLoginOtpEmailRequest): Promise<SendOtpResult> {
+    const user = await this.userService.getUserAsync({ email: sendLoginOtpEmailRequest.email });
+
+    if(!user) throw new UnauthorizedException("User doesn't exist.");
+
+    const sendOtpResult = await this.otpService.createEmailOtpAsync(sendLoginOtpEmailRequest.email);
```
Comment (ismetkizgin) [inline src/login/send-otp-email-login.controller.ts:27]: There is otpValue in the otp result. otpValue should not be sent in the response.

File: src/login/send-otp-email-login.controller.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    if(sendOtpResult.isSent === true) {
+      const otpEmailCreatedEvent = new OtpEmailCreatedEvent();
+      otpEmailCreatedEvent.otpValue = sendOtpResult.otpValue;
+      otpEmailCreatedEvent.emailAddress = sendLoginOtpEmailRequest.email;
+      otpEmailCreatedEvent.authenticationAction = AuthenticationAction.LOGIN;
+      this.eventEmitter.emit("otp.email.created", otpEmailCreatedEvent);
```
Comment (ismetkizgin) [inline src/login/send-otp-email-login.controller.ts:34]: refactor:

```
this.eventEmitter.emit("otp.email.created", {
        otpValue: sendOtpResult.otpValue,
        emailAddress: sendLoginOtpEmailRequest.email,
        authenticationAction: AuthenticationAction.LOGIN,
      });
```

**Review by ismetkizgin — CHANGES_REQUESTED** [review]

File: test/integration/send-otp-email-login.e2e-spec.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+          }).one();
+
+        const responseEmail = await request(app.getHttpServer())
+          .post("/send-login-otp-email")
+          .send(sendLoginOtpEmailRequest)
+          .expect(201);
+
+        expect(responseEmail.body.isSent).toEqual(true);
+        expect(responseEmail.body).toHaveProperty("expiresAt");
+        expect(responseEmail.body).toHaveProperty("otpValue");
```
Comment (ismetkizgin) [inline test/integration/send-otp-email-login.e2e-spec.ts:82]: In the response, there should be no otpValue.

File: src/login/dto/send-login-otp-email-response.dto.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import { ApiProperty } from "@nestjs/swagger";
+
+export class SendLoginOtpEmailResponse {
+  @ApiProperty({
+    required: true,
+  })
+  is_sent: boolean;
```
Comment (ismetkizgin) [inline src/login/dto/send-login-otp-email-response.dto.ts:7]: Let's name it using camelCase naming convention.

File: test/integration/send-otp-email-login.e2e-spec.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+            email: user.email,
+          }).one();
+
+        const responseEmail = await request(app.getHttpServer())
+          .post("/send-login-otp-email")
+          .send(sendLoginOtpEmailRequest)
+          .expect(201);
+
+        expect(responseEmail.body.is_sent).toEqual(true);
+        expect(responseEmail.body).toHaveProperty("expires_at");
```
Comment (ismetkizgin) [inline test/integration/send-otp-email-login.e2e-spec.ts:81]: Could you check if the date is greater than the current time?

File: src/login/dto/send-login-otp-email-response.dto.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+  @ApiProperty({
+    required: true,
+  })
+  expires_at: Date;
```
Comment (ismetkizgin) [inline src/login/dto/send-login-otp-email-response.dto.ts:12]: Let's name it using camelCase naming convention.

**Review by ismetkizgin — CHANGES_REQUESTED** [review]

File: src/login/dto/send-login-otp-email-request.dto.spec.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
+import { SendLoginOtpEmailRequest } from "./send-login-otp-email-request.dto";
+import { faker } from "@faker-js/faker";
+import { validate } from "class-validator";
+
+describe("LoginDto Validation", () => {
```
Comment (mfozmen) [inline src/login/dto/send-login-otp-email-request.dto.spec.ts:5]: SendLoginOtpEmailRequest validation

File: src/login/dto/send-login-otp-email-request.dto.spec.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+  it("should fail validation when LoginDto has empty username and email fields", async () => {
+      const user = new SendLoginOtpEmailRequest();
```
Comment (mfozmen) [inline src/login/dto/send-login-otp-email-request.dto.spec.ts:7]: user? it should be request at least.

File: src/login/dto/send-login-otp-email-request.dto.spec.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+      expect(errors.length).toBe(1);
+      expect(errors[0].constraints).toEqual({
+        isEmail: "email must be an email",
+        isNotEmpty: "email should not be empty",
+      });
+  });
+
+  it("should fail validation when email address is invalid", async () => {
+      const user = new SendLoginOtpEmailRequest();
```
Comment (mfozmen) [inline src/login/dto/send-login-otp-email-request.dto.spec.ts:18]: user? it should be request at least.

File: src/login/dto/send-login-otp-email-response.dto.spec.ts [hunk 11]
```diff
@@ -0,0 +1,10 @@
+import { SendLoginOtpEmailResponse } from "./send-login-otp-email-response.dto";
+
+describe("SendLoginOtpEmailResponse", () => {
+  it("should create a LoginResponse object with optional properties", () => {
```
Comment (mfozmen) [inline src/login/dto/send-login-otp-email-response.dto.spec.ts:4]: LoginResponse -> SendLoginOtpEmailResponse

File: src/login/send-otp-email-login.controller.spec.ts [hunk 12]
```diff
@@ -0,0 +1,159 @@
+import {
```
Comment (mfozmen) [inline src/login/send-otp-email-login.controller.spec.ts:1]: Rename the file as send-login-otp-email.controller.spec.ts

File: src/login/send-otp-email-login.controller.ts [hunk 13]
```diff
@@ -0,0 +1,42 @@
+import { Inject, Controller, Post, Body, UnauthorizedException } from "@nestjs/common";
```
Comment (mfozmen) [inline src/login/send-otp-email-login.controller.ts:1]: Rename the file as send-login-otp-email.controller.ts

File: src/login/send-otp-email-login.controller.spec.ts [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+  import { TokenService } from "../token/token.service";
+  import { UnauthorizedException } from "@nestjs/common";
+  import { UserModule } from "../user/user.module";
+  import { UserService } from "../user/user.service";
+  import { classes } from "@automapper/classes";
+  import { getRepositoryToken } from "@nestjs/typeorm";
+  import {EventEmitterModule} from '@nestjs/event-emitter';
+  import { faker } from "@faker-js/faker";
+  
+  describe("SendLoginOtpEmailController", () => {
```
Comment (mfozmen) [inline src/login/send-otp-email-login.controller.spec.ts:31]: Rename all mock variables better. Define the variable. For example mockUser doesn't define the user. If it's a valid user then rename as validUser or if it's an invalid user rename as invalidUser like that.

File: src/login/send-otp-email-login.controller.ts [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    private readonly otpService: OtpService,
+    private readonly eventEmitter: EventEmitter2
+  ) {}
+
+  @Post("send-login-otp-email")
+  async sendLoginOtpEmailAsync(@Body() sendLoginOtpEmailRequest: SendLoginOtpEmailRequest):
+   Promise<SendLoginOtpEmailResponse> {
+    const user = await this.userService.getUserAsync({ email: sendLoginOtpEmailRequest.email });
+
+    if(!user) throw new UnauthorizedException("User doesn't exist.");
```
Comment (mfozmen) [inline src/login/send-otp-email-login.controller.ts:25]: Remove user doesn't exist message for now. In future this can be handled while sending e-mail.

**Review by mfozmen — CHANGES_REQUESTED** [review]

**Review by mfozmen — APPROVED** [review]

**Pull request merged**